docs/explanations/model_diagnostics/index.rst

```diff
@@ -0,0 +1,45 @@
+
+Next, you should ensure you model has zero degrees of freedom (as best you can); whilst your ultimate goal may be to some an optimization problem with degrees of freedom, you should always start from a square model first. Firstly, this is because many of the model diagnosis tools work best with square models. Secondly, all models are based on the foundation of a square model; an optimization problem is just a square model with some degrees of freedom added. If your underlying square model is not well-posed then any more advanced problem you solve based on it is fundamentally flawed (even if it solves).
+
+Once you have a square model, the next step is to check to see if there are any issues with the structure of the model; for example structural singularities or unit consistency issues. As these issues exist in the very structure of the model, it is possible to check for these before calling a solver and in doing so make it more likely that you will be able to successfully solve the model. Any issues identified at this stage should be resolved before trying to move forwards.
+
+Once you have ensured there are no structural issues with your model, the next step is to try to solve your model (or initialize it if you haven't already) using the solver of your choice (you should also experiment with different solvers if needed). Hopefully you will get some solution back from the solver, even if it is not optimal and/or feasible (in cases where you get critical solver failures, see the section later in this documentation).
+
+Once you have at least a partial solution to your model, the next step is to check for numerical issues in your model, such as possible bounds violations and poor scaling. Due to their nature, numerical issues depend on having a solution to the model, and they can often be limited to certain model states (i.e., it is possible to have a model which behaves well at one model state only to fail badly if you change your model state). As such, numerical checks should be performed at a number of points across the full range of expected states to try to ensure that the model is well-posed across the full modeling range. Any issues that are identified here should be addressed before moving on, and remember that after any changes you should always start by checking for structural issues again.
```

Comment:
I think this explanation (and the accompanying diagram) is really good. However, the explanation is a large amount of text that I think people (myself included) are likely to skip over. It might be nice to separate the paragraphs into titled subsections so people can easily skip to the subsection they are interested in. So that these paragraphs can be sub-sections of "General Workflow", maybe the document should use the format
```rst
==========================
Model Diagnostics Workflow
==========================
...

General Workflow
================
...

Choose a model to debug
-----------------------

Start with a square model
-------------------------

Check for structural issues
---------------------------

Check for numeric issues
------------------------

etc.
```